backend/msgpack/queue.go

```diff
@@ -120,21 +125,43 @@ func (q *queue) drain() {
 				xlog.NewLogField("instance", q.instance.Endpoint()),
 				xlog.NewLogErrField(err),
 			).Error("write data error")
-			q.metrics.connWriteErrors.Inc(1)
 		}
 		buf.Close()
 	}
 	q.conn.Close()
 }
 
+func (q *queue) reportQueueSize(reportInterval time.Duration) {
+	ticker := time.NewTicker(reportInterval)
+	for range ticker.C {
+		q.RLock()
```

Comment:
Hmm, I don't think a done channel would actually protect us from adding a buffer to a closed channel. With a done channel `Enqueue` would look something like:

```
...
select {
case <-q.doneCh:
    // return closed error
default:
    // continue to adding metric 
}
...
```

Then `Close` would look something like:
```
...
close(q.doneCh)
close(q.bufCh)
...
```

But since there's no synchronization between the two channels it's conceivable for one goroutine to check that the done channel isn't closed, another goroutine then closes both the done channel and the buffer channel, and finally the first goroutine tries to put a metric on the buffer channel but panics because it is closed already.